deps/NNNN-observability-metrics.md

```diff
@@ -0,0 +1,261 @@
+* Modifying the core `DistributedRuntime` behaviors
+
+## Requirements
+
+<!-- [Optional - if not applicable omit] List out any additional requirements in numbered subheadings. Requirements should be measurable and will be used to determine if a DEP has been successfully implemented or not. Use all-caps, bolded terms like MUST and SHOULD when describing each requirement. -->
+
+### REQ 1: Automatic Metrics Collection
+
+* **Description:** All component endpoints using the `DistributedRuntime` framework MUST automatically collect basic metrics including request count, duration, bytes transferred, and error counts without requiring any code changes.
+* **Rationale:** Automatic metrics collection ensures consistent observability across all components and reduces the burden on developers to implement basic monitoring.
+* **Measurability:** Verify that all component endpoints automatically expose metrics.
+
+### REQ 2: Standardized Metric Naming
+
+* **Description:** All metrics MUST follow the standard `dynamo_component_*`, `dynamo_frontend_*`, and other `dynamo_<specific-component>_*` prefix in names. Additional **labels** such as `dynamo_namespace`, `dynamo_component`, `dynamo_endpoint` are automatically added to the names.
+* **Rationale:** Consistent naming and labeling enables effective aggregation, filtering, and analysis across the entire Dynamo ecosystem.
+* **Measurability:** Confirm that all metrics follow the established naming patterns and include the required labels.
+
+### REQ 3: Custom Metrics Support
+
+* **Description:** The `MetricsRegistry` MUST support creation of additional custom metric types (counters, gauges, histograms) with proper labeling and integration with the automatic metrics system.
+* **Rationale:** Custom metrics allow components to expose domain-specific observability data while maintaining consistency with the overall metrics framework.
+* **Measurability:** Verify that components can create and use custom metrics through the `MetricsRegistry` interface, implemented in the `DistributedRuntime` framework.
+
+### REQ 4: HTTP Metrics Endpoint
+
+* **Description:** Each process containing one or more components and/or endpoints MUST expose metrics in Prometheus format via HTTP endpoints that can be scraped by monitoring systems.
+* **Rationale:** HTTP endpoints enable integration with standard monitoring tools like Prometheus, Grafana, and other observability platforms.
+* **Measurability:** Confirm that components expose `/metrics` endpoints that return valid Prometheus-formatted data.
+
+### REQ 5: Developer Documentation
+
+* **Description:** Comprehensive documentation MUST be provided for developers on how to use the metrics system, including examples and best practices.
+* **Rationale:** Clear documentation reduces the learning curve and ensures consistent implementation across teams.
+* **Measurability:** Verify the existence and completeness of developer guides, examples, and API documentation.
+
+# Proposal
```

Comment:
Can we add back the implementation proposal / pseudo code? I think the proposal itself here is to add a common metrics registry struct with traits for adding metrics at the different levels .... I think just adding back what you had makes sense